### PR TITLE
Sections and Coteachers

### PR DESCRIPTION
**Pull request by jordan-springer**
# Description

The recent addition of `section_instructors` has allowed us the ability to parse which teachers are: 
* 1. the section owner
* 2. the teacher inviting other teachers
* 3. the coteachers we are otherwise unaccounting for

## Decomposition
* 1. `section_instructors` contains teachers missing from our data

```
select 
    date_trunc('month',sei.created_at)      as created_month,
    use.user_type,
    count(distinct sei.instructor_id)   as num_coteachers

from dashboard_production.section_instructors   as sei 
left join dashboard_production.users            as use 
    on sei.instructor_id = use.id 
    -- and user_type = 'teacher'

where sei.deleted_at is null 
    and use.id is null 

group by 1,2
```

According to this query, there are no ***missing*** teachers, but there are likely ways in which we are over/undercounting the number of teachers teaching sections...

* 2. Are there any ***missing teachers*** in Hydrone?..

*** No, I don't think so ***

```
with 
teachers as (
    select * 
    from analytics.dim_teachers 
    where teacher_id is not null 
),

missing as (
    select instructor_id, section_id
    from analytics.stg_dashboard__section_instructors 
    where deleted_at is null 
), 

final as (
    select *
    from missing
    where instructor_id not in (
        select user_id from teachers) )
        
select * 
from final 
```

* 3. If there are no missing teachers, then how do we need to augment Hydrone to capture coteachers/section owners?
Example `section_id` for reference (Hydrone vs dashboard)
```
select school_year, section_id, is_section_owner, count(distinct teacher_id)
from combined 
where section_id = 5181677
group by 1,2,3

vs 

select
    case when sei.instructor_id = sec.user_id then 1 else 0 end as is_section_owner,
    count(distinct instructor_id)
from dashboard.dashboard_production.section_instructors as sei
left join dashboard.dashboard_production.sections       as sec 
    on sei.section_id = sec.id
where section_id = 5181677
group by 1
```

** a. Model `is_section_owner` flag into `int_section_mapping`
** b. Reference this flag downstream in models involving both teacher and section

## Links

Jira ticket(s): [dataops-548](https://codedotorg.atlassian.net/browse/DATAOPS-548)

## Testing story

- [x] Does your change include appropriate tests on key columns?
      eg.
      - `not_null`
      - `unique`
      - `dbt_utils.unique_combination_of_columns: , ["value","value","value"...]

**Note: when submitting a new model for review please make sure the following have been tested:**

1. The model compiles (`dbt build -m 'your_model'`)
         or: _has the dbt Cloud job succeeded?_
3. The model runs (`dbt run -m 'your_model'`)
4. The model produces accessible data in the DW (`select 1 from 'your_model'`)

## Privacy

- [ ] 1.	Does this change involve the collection, use, or sharing of new Personal Data?
- [x] 2.    Do these data exist in the appropriate schema(s)? 
- [ ] 3.	Does this change involve a new or changed use or sharing of existing Personal Data?
- [ ] 4.    Consider: will this data be visible on Tableau? will this data be surfaced in a report exported from Trevor?
- [ ] 5.    If yes to any of the above, please list the models, columns, and justification below:
      i.
      ii.
      iii. 

## PR Checklist:
**Note: if these are not all checked, the PR will be sent back.**

- [x] Tests provide adequate coverage
- [x] Privacy and Security impacts have been assessed
- [x] Code adheres to [style guide](https://docs.getdbt.com/best-practices/how-we-style/0-how-we-style-our-dbt-projects)👀 and is **[DRY](https://docs.getdbt.com/terms/dry)**
- [x] Code is well-commented (**please do not leave extraneous commentary in model code, if it is for the purpose of documentation, please relocate accordingly)
- [x] Appropriate documentation has been provided (see `.yml.`, did `dbt docs generate` succeed?)
- [ ] New features are translatable or updates will not break up/downstream models
- [ ] Relevant documentation has been added or updated (i.e. `dbt docs` has been updated successfully on [Github Pages](code-dot-org.github.io/analytics/)
- [x] Pull Request is labeled appropriately (eg. `chore/`, `feature/`, `fix/`)
- [ ] Follow-up work items (including potential tech debt) are tracked and linked (if applicable)

### CODE REVIEW

File: dbt/models/intermediate/int_section_mapping.sql [hunk 1]
```diff
@@ -36,13 +36,26 @@ sections as (
     from {{ ref('stg_dashboard__sections') }}
 ),
 
+section_instructors as (
+    select 
+        instructor_id,
+        section_id
+    from {{ ref('stg_dashboard__section_instructors') }}
```
Comment (nataliazm99) [inline dbt/models/intermediate/int_section_mapping.sql:43]: This table (`stg_dashboard__section_instructors`) is dropping records where `invited_by is null` which leads to errors identifying who the section owner is

File: dbt/models/intermediate/int_section_mapping.sql [hunk 2]
```diff
@@ -36,13 +36,26 @@ sections as (
     from {{ ref('stg_dashboard__sections') }}
 ),
 
+section_instructors as (
+    select 
+        instructor_id,
```
Comment (nataliazm99) [inline dbt/models/intermediate/int_section_mapping.sql:41]:
```suggestion
        instructor_id as teacher_id,
```
We should keep consistent labels, either user_id or teacher_id

**Review by nataliazm99 — CHANGES_REQUESTED** [review]
There are two main issues here:
1. Table stg_dashboard__section_instructors is dropping records where invited_by_id is null, which leads to errors in assigning section ownership.

2. These changes are not yet capturing all teachers of a section in `int_section_mapping`, `int_active_sections`, and `dim_sections`, which is the main purpose of this work. As detailed [here](https://codedotorg.atlassian.net/browse/DATAOPS-548?focusedCommentId=27408) and in the comments I had done for [PR 100](https://github.com/code-dot-org/analytics/pull/100).
What we want to have is one record for each section||teacher in each of these tables, including co-teachers and lead teacher, and to have a way to indicate if the teacher is the section owner.

**Comment by jordan-springer** [comment]
Currently testing the following to decide how to define a unique row in `int_section_mapping`
```
with validation_errors as (

    select
        student_id, school_year, section_id, teacher_id -- , is_section_owner
    from "dev"."dbt_jordan"."int_section_mapping"
    group by student_id, school_year, section_id, teacher_id --, is_section_owner
    having count(*) > 1

)

select *
from validation_errors
order by student_id
```

**Pull request closed without merging**